src/main/scala/faunadb/importer/persistence/ConnectionPool.scala

```diff
@@ -5,94 +5,42 @@ import com.ning.http.client._
 import faunadb._
 import faunadb.importer.report._
 import java.util.concurrent.TimeUnit._
-import java.util.concurrent.atomic._
-import scala.annotation.tailrec
 import scala.collection.JavaConverters._
 
 trait ConnectionPool {
-  def borrowClient(): FaunaClient
-  def returnClient(client: FaunaClient): Unit
-  def maxConcurrentReferences: Int
+  def pickClient(): FaunaClient
   def close(): Unit
 }
 
 object ConnectionPool {
-  def apply(endpoints: Seq[String], secret: String, maxRefCountPerEndpoint: Int): ConnectionPool =
-    new FaunaClientPool(endpoints, secret, maxRefCountPerEndpoint)
-}
-
-private final class FaunaClientPool(endpoints: Seq[String], secret: String, maxRefCountPerEndpoint: Int)
-  extends ConnectionPool {
-
-  private val refCountByClient: Map[FaunaClient, AtomicInteger] = {
-    def newClient(endpoint: String = null): FaunaClient =
-      FaunaClient(secret, endpoint, httpClient = new HttpWrapper())
+  def apply(endpoints: Seq[String], secret: String): ConnectionPool = {
+    val httpWrapper = new HttpWrapper // Reuse http client
 
-    val clients =
-      if (endpoints.isEmpty) Seq(newClient())
-      else endpoints.map(newClient)
-
-    clients
-      .map(_ -> new AtomicInteger(0))
-      .toMap
+    endpoints match {
+      case endpoint :: Nil => new SingleConnPool(FaunaClient(secret, endpoint, httpClient = httpWrapper))
+      case Nil             => new SingleConnPool(FaunaClient(secret, httpClient = httpWrapper))
+      case _               => new MultipleConnPool(endpoints map (FaunaClient(secret, _, httpClient = httpWrapper)))
+    }
   }
+}
 
-  private val clientsByIndex = refCountByClient.toIndexedSeq
-  private val poolSize = refCountByClient.size
-
-  val maxConcurrentReferences: Int = maxRefCountPerEndpoint * poolSize
+private final class SingleConnPool(client: FaunaClient) extends ConnectionPool {
+  def pickClient(): FaunaClient = client
+  def close(): Unit = client.close()
+}
 
+private final class MultipleConnPool(clients: Seq[FaunaClient]) extends ConnectionPool {
   @volatile
-  private var searchIndex = -1
+  private[this] var searchIndex = -1
+  private[this] val clientsByIndex = clients.toIndexedSeq
+  private[this] val poolSize = clients.size
 
-  @tailrec
-  def borrowClient(): FaunaClient = {
-    var pickedClient: FaunaClient = null
-    var pickedRefCount: AtomicInteger = null
-    var minRefCountFound = Int.MaxValue
-
-    // Start from a different position everytime so we don't
-    // prefer the first few clients in the pool
+  def pickClient(): FaunaClient = {
     searchIndex = Math.max((searchIndex + 1) % poolSize, 0)
```

Comment:
if you start `searchIndex` with 0 you won't need `Math.max`, the only thing is that you would have to increment the variable after get the client, so a temp variable would be needed

Reply:
also I saw `searchIndex` is volatile, is this method used concurrently? if yes, shouldn't this be an atomic operation?

Reply:
Good call about the `Math.max`.

This method is used concurrently but we don't care about atomicity. This is a rough distributions, we don't care if two threads get the same client.

Reply:
If there’s lots of concurrent access, you’re probably better off with a random distribution. W/o atomic access or synchronized you’re almost guaranteed that two threads racing are going to get the same client.

Reply:
I've tested it with 50 threads using 3 clients. It seemed fine when I looked at datadog. Even when threads get the same client, during the corse of a long run them balance out.